### DISCUSSION TITLE
VFS plugins

### DISCUSSION THREAD

**Discussion by Smithx10** [drakkan/sftpgo#553] [Ideas]
Good day,

I was thinking about this today, and I believe it would be worth while to implement a plugin interface for a VFS.  This would allow for users to iterate on the backend VFS implementations without changing the sftpgo binary.     

Thoughts?

**Comment by drakkan** [comment]
Hi,

yes VFS plugins would be a useful addition but it is not so easy. We should dynamically register configurations to expose them in REST API and web admin UI, and I think there are other rough edges.

If you have time you could take a look, a well written patch would be welcome, I have several features to add and I think I could analyze this myself after releasing v2.2.0

**Comment by atruskie** [comment]
I also thought this would be useful. 

Scenario: user upload endpoints exposed by sftpgo, then data is transformed/organized/renamed. However, I'd like to expose a subset of those files with different names for download later on.

I hesitated to suggest it since it could be arbitrarily complex to implement.

However, there are various parts of this that could be useful:

- a file system mapping (useful for exposing a different view of data backed by a "real" provider)
- a read only VFS (reduced scope)
- a VFS REST schema that sftpgo could query

The common theme here I think is "capabilities". I only need a few capabilities for a VFS to enable dozens of new scenarios. If a VFS implementation could support subsets of features it might make it easier to work with.
